### PR TITLE
C-variadic error improvements

### PR DESCRIPTION
**Pull request by nicholasbishop**
A couple improvements for c-variadic errors:

1. Fix the bad-c-variadic error being emitted multiple times. If a function incorrectly contains multiple `...` args, and is also not foreign or `unsafe extern "C"`, only emit the latter error once rather than once per `...`.

2. Explicitly reject `const` C-variadic functions. Trying to use C-variadics in a const function would previously fail with an error like "destructor of `VaListImpl<'_>` cannot be evaluated at compile-time". Add an explicit check for const C-variadics to provide a clearer error: "functions cannot be both `const` and C-variadic". This also addresses one of the concerns in https://github.com/rust-lang/rust/issues/44930: "Ensure that even when this gets stabilized for regular functions, it is still rejected on const fn."

### CODE REVIEW

File: compiler/rustc_ast_passes/src/ast_validation.rs [hunk 1]
```diff
@@ -482,9 +482,32 @@ impl<'a> AstValidator<'a> {
...
+    ///
+    /// C-variadics must be:
+    /// - Non-const
+    /// - Either foreign, or free and `unsafe extern "C"` semantically
     fn check_c_variadic_type(&self, fk: FnKind<'a>) {
+        let Some(variadic_span) = fk
+            .decl()
+            .inputs
+            .iter()
+            .find(|arg| matches!(arg.ty.kind, TyKind::CVarArgs))
```
Comment (oli-obk) [inline compiler/rustc_ast_passes/src/ast_validation.rs:495]: you could also collect all of them into a `vec` and make the variadic label point to all of them at once.

**Review by oli-obk — APPROVED** [review]
lgtm. let me know about your opionion on making the diagnostic point to all the variadics at once

File: compiler/rustc_ast_passes/src/ast_validation.rs (hunk 1 above)
Comment (nicholasbishop) [reply compiler/rustc_ast_passes/src/ast_validation.rs]: Updated the PR to collect into a vec as suggested

**Comment by oli-obk** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**